apps/system/js/audio_channel_manager.js

```diff
@@ -0,0 +1,267 @@
+      this._smallerVolumeChannels = [];
+    },
+
+    /**
+     * Add an app into manager to manage its audio channels.
+     * @param {MozBrowser} app An app you want to manage.
+     */
+    register: function(app) {
+      app.allowedAudioChannels.forEach(function(channel) {
+        channel.onactivestatechanged = function(evt) {
+          var channel = evt.target;
+          if (channel.isActive()) {
+            this._handleAudioChannelCompeting(channel);
+          } else {
+            this._resumeLatestPausedChannel();
+            this._setVolume(this._smallerVolumeChannels, 1);
+          }
+        };
+      });
+    },
+
+    /**
+     * Remove an app from manager,
+     * if you don't wnat to manage it anymore.
+     * @param {MozBrowser} app An app you want to remove.
+     */
+    unregister: function(app) {
+      app.allowedAudioChannels.forEach(function(channel) {
+        channel.onactivestatechanged = undefined;
+      });
+    },
+
+    /**
+     * Handel the new audio channel.
+     * @param {BrowserElementAudioChannel} channel 
+     * A New channel you want to handel.
+     */
+    _handleAudioChannelCompeting: function(channel) {
+      var currentChannel;
+      switch (this.compete(channel.name)) {
+        case PLAY_CURRENT_AND_NEW_CHANNEL:
+          this._play(channel);
+          this._currentChannels.push(channel);
+          break;
+
+        case PAUSE_CURRENT_CHANNEL:
+          currentChannel = this._currentChannels.pop();
+          this._pause(currentChannel);
+          break;
+
+        case PAUSE_THEN_RESUME_CURRENT_CHANNEL:
+          currentChannel = this._currentChannels.pop();
+          this._pause(currentChannel);
+          this._pausedApps.push(currentChannel);
+          break;
+
+        case VOLUME_DOWN_CURRENT_CHANNEL:
+          this._currentChannels.forEach(function(channel) {
+            this._play(channel, 0.2);
+            this._smallerVolumeChannels.push(channel);
+          }.bind(this));
+          break;
+
+        case VOLUME_DOWN_NEW_CHANNEL:
+          this._play(channel, 0.2);
+          this._smallerVolumeChannels.push(channel);
+          break;
+
+        case PAUSE_THEN_VIBRATE_CURRENT_CHANNEL:
+          currentChannel = this._currentChannels.pop();
+          this._pause(currentChannel);
+          this._vibrate();
+          break;
+
+        case VIBRATE_NEW_CHANNEL:
+          this._vibrate();
+          break;
+      }
+    },
+
+    /**
+     * Compete new and playing audio channels.
+     * The policy is listed in page 12 in [1].
+     * [1]: https://bug961967.bugzilla.mozilla.org/attachment.cgi?id=8541542.
+     * @param {String} newAudioChannel Channel name.
+     * @return {Number} The competing result.
+     */
+    _compete: function(newAudioChannel) {
+      var COMPETING_POLICIES = {
+        'normal': {
+          'normal': PAUSE_CURRENT_CHANNEL,
+          'foregroundContent': PAUSE_CURRENT_CHANNEL,
+          'backgroundContent': -1,
```

Comment:
Where does foregroundContent & backgroundContent come from? Gecko should not tell this.